cpp/include/cudf/detail/null_mask.cuh

```diff
@@ -199,253 +214,249 @@ __global__ void subtract_set_bits_range_boundaries_kernel(bitmask_type const* bi
+                                                           count_bits_policy count_bits,
+                                                           rmm::cuda_stream_view stream)
 {
-  size_t const num_indices = std::distance(indices_begin, indices_end);
+  rmm::device_uvector<size_type> d_bit_counts(num_ranges, stream);
 
-  CUDF_EXPECTS(num_indices % 2 == 0, "Array of indices needs to have an even number of elements.");
-  for (size_t i = 0; i < num_indices / 2; i++) {
-    auto begin = indices_begin[i * 2];
-    auto end   = indices_begin[i * 2 + 1];
-    CUDF_EXPECTS(begin >= 0, "Starting index cannot be negative.");
-    CUDF_EXPECTS(end >= begin, "End index cannot be smaller than the starting index.");
-  }
-
-  if (num_indices == 0) {
-    return std::vector<size_type>{};
-  } else if (bitmask == nullptr) {
-    std::vector<size_type> ret(num_indices / 2);
-    for (size_t i = 0; i < num_indices / 2; i++) {
-      ret[i] = indices_begin[2 * i + 1] - indices_begin[2 * i];
-    }
-    return ret;
-  }
-
-  size_type num_ranges = num_indices / 2;
-  std::vector<size_type> h_first_indices(num_ranges);
-  std::vector<size_type> h_last_indices(num_ranges);
-  thrust::stable_partition_copy(thrust::seq,
-                                indices_begin,
-                                indices_end,
-                                thrust::make_counting_iterator(0),
-                                h_first_indices.begin(),
-                                h_last_indices.begin(),
-                                [](auto i) { return (i % 2) == 0; });
-
-  auto d_first_indices = make_device_uvector_async(h_first_indices, stream);
-  auto d_last_indices  = make_device_uvector_async(h_last_indices, stream);
-  rmm::device_uvector<size_type> d_null_counts(num_ranges, stream);
-
-  auto word_num_set_bits  = thrust::make_transform_iterator(thrust::make_counting_iterator(0),
-                                                           word_num_set_bits_functor{bitmask});
-  auto first_word_indices = thrust::make_transform_iterator(
-    thrust::make_counting_iterator(0),
-    // We cannot use lambda as cub::DeviceSegmentedReduce::Sum() requires
-    // first_word_indices and last_word_indices to have the same type.
-    to_word_index(true, d_first_indices.data()));
-  auto last_word_indices = thrust::make_transform_iterator(
-    thrust::make_counting_iterator(0),
-    // We cannot use lambda as cub::DeviceSegmentedReduce::Sum() requires
-    // first_word_indices and last_word_indices to have the same type.
-    to_word_index(false, d_last_indices.data()));
-
-  // first allocate temporary memory
+  auto num_set_bits_in_word = thrust::make_transform_iterator(thrust::make_counting_iterator(0),
+                                                              count_set_bits_in_word{bitmask});
+  auto first_word_indices =
+    thrust::make_transform_iterator(first_bit_indices, bit_to_word_index{false});
```

Comment:
It is satisfying to see how well the code reads after the refactor. Kudos!

Would there be value in switching the bool to an enum?
```c++
auto first_word_indices = thrust::make_transform_iterator(first_bit_indices, bit_to_word_index{INCLUSIVE});
auto last_word_indices  = thrust::make_transform_iterator(last_bit_indices,  bit_to_word_index{EXCLUSIVE});
```

Reply:
I didn't think the increase in legibility was worth adding another enum, so I opted for brevity here. Comparing this case to the "count bits policy" enum, this case is much more self-contained (the entire struct is <10 lines so the influence of the parameter is localized) while the count bits policy is propagated through multiple functions and more dramatically alters behavior. We can add an enum if you still think it is valuable after reading this comment, I just wanted to make the argument for why I used an enum in one case but not the other.